### PR TITLE
Match OpenLR LinearReferences to edges

### PR DESCRIPTION
**Pull request by danpat**
# Issue

This (work-in-progress) PR adds the ability to match OpenLR Line Location References to Valhalla edges.  The goal is to support ingestion of OpenLR described traffic data.  It addresses https://github.com/valhalla/valhalla/issues/1224

This PR is a work-in-progress.

## Tasklist

 - [ ] Determine proper graph level when performing loki nearest edge search
 - [ ] Add support for first-coordinate exit-bearing filtering (only last-coordinate incoming bearing is currently filtered)
 - [ ] Filter edges by FOW/FRC when doing shortest-path search
 - [ ] How to continue searching for candidate paths when shortest path is too short?
 - [ ] Support LR offset values (and return % offsets in `EdgeMatch` results) (low priority)
 - [ ] Find k-shortest paths, and rank, returning best match (OpenLR toolkit does this)

 - [ ] add tests
 - [ ] review - you must request approval to merge any PR to master
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Generally use squash merge to rebase and clean comments before merging
 - [ ] Add to release notes in valhalla-docs
 - [ ] update relevant documentation if this is an API impacting change

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

**Comment by dnesbitt61** [comment]
I don't think we want a valhalla_associate_osmlr process here - I think we want methods/class to accept an OSMLR descriptor and return the edge(s) that associate. The association code should live in baldr (where the graph access code generally lives). OpenLR perhaps should also live in baldr since midgard is generally low level math code and utilities.

File: valhalla/midgard/openlr.h [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    byte unused_1 : 1;
+    bool ArF1 : 1;
+    bool no_point : 1;
+    bool ArF0 : 1;
+    bool has_attributes : 1;
+    byte version  : 3;
+
+    // Note: wire-format is big-endian
+    // Bytes 1-6
+    std::int32_t fixed_longitude : 24;
```
Comment (dnesbitt61) [inline valhalla/midgard/openlr.h]: Vahalla uses uint8_t, int16_t, uint64_t etc. throughout the code base. No real reason to prefix with std.

**Comment by oxidase** [comment]
@danpat please could you review the pull request?

*  `location_referencer` code is removed from the change set. `location_referencer` will be a part of a separate repository and can be back-ported into Valhalla code base later
* OpenLR packing/unpacking uses explicit bit operations because bit-fields packing is implementation-dependent

**Comment by danpat** [comment]
@oxidase 👍 looks great!  Moving away from a platform-dependent binary decoding is a great idea, and will make it easier to support the OpenLR XML form if we need to for more precision.

**Comment by dnesbitt61** [comment]
XML? I think we should use protocol buffer if we do not feel that a compact binary representation is needed. We also shouldn't make that decision until we know whether we need to decode/match on the client side or whether all that work will be done server side (thinking about dynamic data like real-time speed, incidents, etc.)

**Comment by danpat** [comment]
@dnesbitt61 I said "if we need it" - and mostly only because XML is the other canonical encoding in the OpenLR standard.  Many existing third parties publish descriptors in that form, so it seems like it might come up at some point in the future.  Not suggesting we support it right now.

**Pull request merged**